test functionality of gold linker on full set of link libraries

Due to a recent post on the mailing lists,
https://groups.google.com/forum/#!topic/dealii/JxUOyd_4eFM
I checked the behavior of OpenMPI on Ubuntu 16.04 on the following trivial test file `test.cc`:
```
int main() {}
```
I tried to compile this with the settings detected by the deal.II configuration and it gives me a linker error as reported by Pete Griffin. The important part is the gold linker:
```
$ mpicxx -fuse-ld=gold test.cc
/usr/lib/openmpi/lib/libmpi_cxx.so: error: undefined reference to 'opal_list_item_t_class'
/usr/lib/openmpi/lib/libmpi_cxx.so: error: undefined reference to 'opal_class_initialize'
collect2: error: ld returned 1 exit status
```
The file links fine the the bfd linker (i.e., without the `-fuse-ld=gold` flag). While this clearly is a bug in the OpenMPI implementation shipped by Ubuntu, I think that our build system is to blame as well. We use a non-standard tool (the gold linker) and should ensure that it really works with the ***final*** configuration selected by the user. As far as I can tell, we need to put the test for the gold linker not early in the configuration process as we currently do (`cmake/checks/check_01_compiler_features.cmake`), but check with the final set of link directories after detection of all third party libraries. @tamiko do you have another suggestion?

--- a/cmake/modules/FindMPI.cmake
+++ b/cmake/modules/FindMPI.cmake
@@ -124,9 +124,48 @@ IF(NOT DEFINED MPI_VERSION OR MPI_VERSION STREQUAL ".")
   SET(MPI_VERSION_MINOR "0")
 ENDIF()
 
+#
+# Make sure that we do not run into underlinking on Debian/Ubuntu systems with
+# lld / ld.gold and missing libopen-pal.so on the link line:
+#
+
+CHECK_COMPILER_SETUP(
+  "${DEAL_II_CXX_FLAGS_SAVED} ${DEAL_II_CXX_FLAGS}"
+  "${DEAL_II_LINKER_FLAGS_SAVED} ${DEAL_II_LINKER_FLAGS}"
+  MPI_UNDERLINKAGE_OK
+  ${MPI_CXX_LIBRARIES} ${MPI_Fortran_LIBRARIES} ${MPI_C_LIBRARIES}
+  )
+
+IF(NOT MPI_UNDERLINKAGE_OK)
+  MESSAGE(STATUS "Trying to avoid underlinkage by expliclitly adding libopen-pal to link line")
+
+  LIST(GET MPI_CXX_LIBRARIES 0 _lib)
+  GET_FILENAME_COMPONENT(_hint ${_lib} DIRECTORY)
+  DEAL_II_FIND_LIBRARY(_mpi_libopen_pal_library
+    NAMES open-pal
+    HINTS ${_hint}
+    NO_DEFAULT_PATH
+    NO_CMAKE_ENVIRONMENT_PATH
+    NO_CMAKE_PATH
+    NO_SYSTEM_ENVIRONMENT_PATH
+    NO_CMAKE_SYSTEM_PATH
+    NO_CMAKE_FIND_ROOT_PATH
+    )
+
+  #
+  # Note: We don't need to check whether the find library call is
+  # successful: If libopen-pal cannot be found then the
+  # DEAL_II_PACKAGE_HANDLE will drop the library automatically.
+  #
+  # In this case the sanity check in cmake/setup_finalize.cmake will fail
+  # and we start dropping -fuse-ld=lld and -fuse-ld=ld.gold from the
+  # command line.
+  #
+ENDIF()
+
 DEAL_II_PACKAGE_HANDLE(MPI
   LIBRARIES
-    OPTIONAL MPI_CXX_LIBRARIES MPI_Fortran_LIBRARIES MPI_C_LIBRARIES
+    OPTIONAL MPI_CXX_LIBRARIES MPI_Fortran_LIBRARIES MPI_C_LIBRARIES _mpi_libopen_pal_library
   INCLUDE_DIRS
     OPTIONAL MPI_CXX_INCLUDE_PATH MPI_C_INCLUDE_PATH
   USER_INCLUDE_DIRS
@@ -145,4 +184,3 @@ DEAL_II_PACKAGE_HANDLE(MPI
     MPI_MPI_H
     MPI_HAVE_MPI_SEEK_SET
   )
-

